services/admin/src/admin/routes/obsolete_cache.py

```diff
@@ -0,0 +1,140 @@
+from libcommon.viewer_utils.asset import delete_asset_dir
+from starlette.requests import Request
+from starlette.responses import Response
+
+from admin.authentication import auth_check
+from admin.utils import (
+    Endpoint,
+    UnexpectedError,
+    get_json_admin_error_response,
+    get_json_ok_response,
+)
+
+DATASET_CONFIG_NAMES_KIND = "dataset-config-names"
+MINIMUM_SUPPORTED_DATASETS = 20_000
+
+
+class DatasetCacheReport(TypedDict):
+    dataset: str
+    cache_records: int
+
+
+def get_supported_dataset_names(
+    hf_endpoint: str,
+    hf_token: Optional[str] = None,
+) -> set[str]:
+    supported_dataset_infos = get_supported_dataset_infos(hf_endpoint=hf_endpoint, hf_token=hf_token)
+    return {dataset_info.id for dataset_info in supported_dataset_infos}
+
+
+def get_obsolete_cache(
+    hf_endpoint: str,
+    hf_token: Optional[str] = None,
+) -> list[DatasetCacheReport]:
+    supported_dataset_names = get_supported_dataset_names(hf_endpoint=hf_endpoint, hf_token=hf_token)
+    existing_datasets = get_all_datasets(DATASET_CONFIG_NAMES_KIND)
+    datasets_to_delete = existing_datasets.difference(supported_dataset_names)
+    return [
+        DatasetCacheReport(dataset=dataset, cache_records=get_cache_count_for_dataset(dataset=dataset))
+        for dataset in datasets_to_delete
+    ]
+
+
+def create_get_obsolete_cache_endpoint(
+    hf_endpoint: str,
+    max_age: int,
+    hf_token: Optional[str] = None,
+    external_auth_url: Optional[str] = None,
+    organization: Optional[str] = None,
+    hf_timeout_seconds: Optional[float] = None,
+) -> Endpoint:
+    async def get_obsolete_cache_endpoint(request: Request) -> Response:
+        try:
+            logging.info("/obsolete-cache")
+            auth_check(
+                external_auth_url=external_auth_url,
+                request=request,
+                organization=organization,
+                hf_timeout_seconds=hf_timeout_seconds,
+            )
+            return get_json_ok_response(
+                get_obsolete_cache(hf_endpoint=hf_endpoint, hf_token=hf_token), max_age=max_age
+            )
+        except Exception as e:
+            return get_json_admin_error_response(UnexpectedError("Unexpected error.", e), max_age=max_age)
+
+    return get_obsolete_cache_endpoint
+
+
+def delete_obsolete_cache(
+    hf_endpoint: str,
+    assets_directory: StrPath,
+    cached_assets_directory: StrPath,
+    hf_token: Optional[str] = None,
+) -> tuple[int, int]:
+    supported_dataset_names = get_supported_dataset_names(hf_endpoint=hf_endpoint, hf_token=hf_token)
+    if len(supported_dataset_names) < MINIMUM_SUPPORTED_DATASETS:
+        raise UnexpectedError(f"only {len(supported_dataset_names)} datasets were found")
+
+    existing_datasets = get_all_datasets(DATASET_CONFIG_NAMES_KIND)
+    datasets_to_delete = existing_datasets.difference(supported_dataset_names)
+    deleted_cache_records = 0
+
+    for dataset in datasets_to_delete:
```

Comment:
(nit) Maybe add a log.info that mentions the dataset being deleted ?

Reply:
and maybe even return datasets_to_delete, instead of just the number -> to show in the admin UI

Reply:
Ok, I kept the same response format as GET /obsolete-cache to have detailed information about which datasets and how many cached records were deleted for each of them.